Cards metadata to vue

chosta opened the pull request:
Continuing the battle with HTML_data.ts:
+ added `CardMetadata` interface (still more attributes to come)
+ added CardRequirements to metadata
+ abstracting the `<div class="content">` by checking if metadata exists. If yes => use Vue components , no => use the HTML_data.ts content
+ first card to be set free is also the best one in the game ->>> `Dust Seals` !

This is WIP so, please just comment on what you think needs to change in the existing code and not what is still missing :)

kberg commented:
I hate to correct you, but the best card in the game is the one that I play to take six plants away from **you**! BAHAHAHAHAHAHA,

Lemme look at your PR.

kberg commented on src/cards/CardRequirement.ts:
```diff
@@ -0,0 +1,28 @@
+        this.isMax = isMax;
+    }
+
+    private getParsedAmount(): string {
```
Include degree for temperature?

Also toString() is a more than adequate name.

kberg commented on src/cards/CardRequirement.ts:
```diff
@@ -0,0 +1,28 @@
+        }
+    }
+
+    public getReqText(): string {
```
Sorry, what does getReqText mean?

chosta replied:
:) means a lazy dev does things twice

kberg replied:
Sorry, I mean requirement or request? Would you mind spelling it out? (Then I won't ask for a comment! ;))

chosta replied:
The curse of knowledge strikes again :)
I'm doing some refactoring, will tag you as soon as I have something complete.

kberg commented on src/cards/DustSeals.ts:
```diff
@@ -1,23 +1,37 @@
+    public metadata: CardMetadata = {
+        description: "Requires 3 or less ocean tiles.",
+        cardNumber: 119,
+        requirements: new CardRequirements([
```
Have CardRequirements accept `...CardRequirement`

Also this can be made much simpler vis. LogBuilder.

Imagine this Api:

```
requirements: CardRequirements.create(b => {
  b.oceans(3).max();
  b.science(12);
});
```

kberg replied:
It's about a hundred of these, that'll be much easier to read.

kberg commented on src/cards/DustSeals.ts:
```diff
@@ -1,23 +1,37 @@
     public name: CardName = CardName.DUST_SEALS;
     public canPlay(player: Player, game: Game): boolean {
-      return game.board.getOceansOnBoard() <= 3 + player.getRequirementsBonus(game);
+        return (
```
Can you disable the auto formatter so reviews are more focused?

chosta replied:
I'd like to give another approach a try instead. First format all the files I want to work on and send as a separate commit. I won't give up on formatting that easy :)
+ I increased default width to 160 columns so let's see how it goes
I appreciate your patience

kberg replied:
> I appreciate your patience
I felt the same way, I did feel a little bad making the request.

chosta replied:
But this is a legitimate concern. We have to appreciate the time and effort of reviewers, so lets find a common ground.

kberg replied:
This is not the thread for it, but so you know, I think the right approach is that we move quickly to adopt a common standard -- if we can reach consensus.

kberg commented on src/cards/IProjectCard.ts:
```diff
@@ -14,4 +15,5 @@ export interface IProjectCard extends ICard {
     hasRequirements?: boolean;
```
I think this makes hasRequirements obsolete, right?

chosta replied:
It does but I dare not make breaking changes at the moment. I'll add it as a to-do and resolve 
after all cards are gone from the HTML_data

kberg replied:
How do you track your TODOs? Are you creating a tracking bug or a comment here, or just do it later?

kberg commented:
Terrific, I love this. Keep it up and to show my thanks I'll pass you `Sabotage` in the draft.

bafolts commented on src/cards/CardMetadata.ts:
```diff
@@ -0,0 +1,8 @@
+    cardNumber: number;
+    description?: string;
+    requirements?: CardRequirements;
+    points?: number; //#TODO -> class to handle points per tag and other special cases
```
I would use `victoryPoints` for this name as it is more specific than `points`.

chosta replied:
good point (pun intended)

bafolts commented on src/cards/CardRequirement.ts:
```diff
@@ -0,0 +1,30 @@
+export class CardRequirement {
+    type: RequirementType;
+    amount: number;
+    isMax?: boolean = false;
```
What does it represent if `isMax` is `undefined` versus `false`?

chosta replied:
You already know that I'm scared of `undefined` and try to avoid it. If I understand correctly you are suggesting that isMax could simply have the ? and lose the default `false` value?

kberg replied:
Or not have ? at all.

bafolts commented on src/cards/CardRequirements.ts:
```diff
@@ -0,0 +1,16 @@
+        return reqTexts.join(" ");
+    }
+    public hasMax(): boolean {
+        return this.requirements.filter((req) => req.isMax).length > 0;
```
This is a good use case for `Array.prototype.some`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some

chosta replied:
yes!

bafolts commented on src/cards/DustSeals.ts:
```diff
@@ -1,23 +1,32 @@
-      return 1;
+        return 1;
     }
+    public metadata: CardMetadata = {
```
Could `tags` and `cardType` also go on `CardMetadata`? We don't have a `SerializedCard` but eventually `SerializedCard` would only need to store the `CardName` and any resources on the card or other dynamic properties.

chosta commented on src/cards/CardRequirements.ts:
```diff
@@ -6,6 +6,15 @@ export class CardRequirements {
         this.requirements = reqs;
     }
 
+    public static create(f: (r: CardRequirement) => void): CardRequirements {
```
@kberg I need some help here :/
Managed to make the first part of the API that you suggested but not sure how to accept multiple CardRequirements in the create method like you suggested. This concept is new to me, feel free to explain what happens in details :)
I'm trying to do that:
`CardRequirements.create((b) => {
            b.oceans(3).max();
            b.science(1)
        }),`

chosta commented:
@kberg followed your guidance for the CardRequirement refactor. I am also ditching the formatter, just fixing files that are having 2 spaces and having it turned off most of the time.
@bafolts I see that there is some code / logic repetition in the cards like victory points, tags, etc. I'd rather focus on having the complete card description using the metadata and then start refactoring and removing the repetitive logic.
Let me know what you think.

kberg commented:
Is this ready for re-review? Sorry for losing track. Also note merge conflict.

chosta commented:
@kberg I got this after merging and not sure how to approach it:
```/home/dev/terraforming-mars/dist/src/Dealer.js:27
        found = deck.projectCards.findByCardName(cardName);
                     ^

TypeError: Cannot read property 'projectCards' of undefined
    at /home/dev/terraforming-mars/dist/src/Dealer.js:27:22
    at Array.forEach (<anonymous>)
    at getProjectCardByName (/home/dev/terraforming-mars/dist/src/Dealer.js:23:19)
    at /home/dev/terraforming-mars/dist/src/Dealer.js:157:20
    at Array.map (<anonymous>)
    at Dealer.loadFromJSON (/home/dev/terraforming-mars/dist/src/Dealer.js:156:28)
    at Game.loadFromJSON (/home/dev/terraforming-mars/dist/src/Game.js:1345:30)
    at Statement.<anonymous> (/home/dev/terraforming-mars/dist/src/database/SQLite.js:65:18)
```

kberg commented:
> @kberg I got this after merging and not sure how to approach it:
> 
> ```
>         found = deck.projectCards.findByCardName(cardName);
>                      ^
> 
> TypeError: Cannot read property 'projectCards' of undefined
>     at /home/dev/terraforming-mars/dist/src/Dealer.js:27:22
>     at Array.forEach (<anonymous>)
>     at getProjectCardByName (/home/dev/terraforming-mars/dist/src/Dealer.js:23:19)
>     at /home/dev/terraforming-mars/dist/src/Dealer.js:157:20
>     at Array.map (<anonymous>)
>     at Dealer.loadFromJSON (/home/dev/terraforming-mars/dist/src/Dealer.js:156:28)
>     at Game.loadFromJSON (/home/dev/terraforming-mars/dist/src/Game.js:1345:30)
>     at Statement.<anonymous> (/home/dev/terraforming-mars/dist/src/database/SQLite.js:65:18)
> ```
> 
> It's obviously an old game that doesn't load properly on my local machine. I could purge the DB but this needs attention because it might break other instances (maybe not the demo).

Confirmed that a fresh DB doesn't exhibit this behavior. As I mentioned in chat, send me the DB and I'll see exactly why.

kberg commented on src/cards/AdvancedEcosystems.ts:
```diff
@@ -1,21 +1,35 @@
+        description: "Requires a Plant tag, a Microbe tag, and an Animal tag.",
+        cardNumber: "135",
+        requirements: new CardRequirements([
+            CardRequirement.tag(Tags.PLANT, -1),
```
Why is this -1? Shouldn't it be 1?

chosta replied:
So the thing here is that we sometimes want the actual "1" in front of the requirement:
`max 1 colony`
And in the case we just want the tag name:
`plant microbe animal`
So I figured we could use the `-1` to skip numbers. Maybe an inline comment would help here. I don't wan to overcomplicate things anyway.

kberg replied:
Could have default be 0 or -1?

kberg commented on src/cards/IProjectCard.ts:
```diff
@@ -14,4 +15,5 @@ export interface IProjectCard extends ICard {
     hasRequirements?: boolean;
     bonusResource?: Resources | undefined;
     bonusMc?: number | undefined;
+    metadata?: CardMetadata;
```
IProjectCard extends ICard, for which metadata is a field. Can this be removed?

chosta replied:
Strange, I remember TS complained about not having it. Now that I removed it its all good.

kberg reviewed (changes requested):
Not bad!

chosta commented on src/cards/CardRequirement.ts:
```diff
@@ -0,0 +1,163 @@
+        }
+    }
+     
+    public getLabel(): string {
```
I prefer dealing with whitespaces by using arr.join(" ") but I'll indulge you here:
The `+=` should help improve it a bit more but this `"max "` is a thorn in my eyes just FYI :)
```
let result: string = this.isMax ? "max " : "";
        const amount = this.amountToString();
        if(amount !== ""){
            result += amount;
            result += " ";
        }
        result += this.parseType();

        return result;
```

chosta commented on src/cards/CardRequirement.ts:
```diff
@@ -0,0 +1,163 @@
+    }
+
+    public getTypePlural(): string {
+        if(this.type === RequirementType.CITIES) {
```
I'll add it to a commit in the future.

chosta commented on src/cards/BreathingFilters.ts:
```diff
@@ -1,23 +1,31 @@
+import { CardRequirement } from "../cards/CardRequirement";
 
 export class BreathingFilters implements IProjectCard {
     public cost: number = 11;
```
I'd like to do that too. No complaints from TS so we should be good to go.

chosta commented:
Alright. Progress made.
+ Did changes from your latest reviews but some items will remain with a TODO status
+ I did a thing where `onPlay` in metadata should be able to describe every remaining element from any card design
+ I figured all elements could be described with `CardRow` and items inside of them so I tried to do something that accepts arrays of different types of elements. It does the job of ordering the elements too. I know it might be a bit controversial but the resulting api allows for a fast and seamless card content implementation.
+ A `CardProductionBox` can already reuse some of the components and classes and can be a part of a `CardRow` so this went well, although I have a feeling we can clean up the code (vue especially) and make it even more reusable.
+ I am slowly moving to a new cards_v2.less file that improves css significantly, especially when it comes to centering items and reducing hardcoded styles (thanks flexbox! <3)
+ Started replacing some of the cards and boy is this satisfying. Everything centered and pixel perfect :). Thanks you @kberg for helping me with these APIs, I like how easy is to describe things in human readable manner. With the current functionality should you approve it, I can cover 80% of the cards.

--- still to come ---
+ unit tests once we are decided on the class and interfaces structure
+ `CardEffect`
+ `CardAction`
+ corporations will be left for last but nothing can stop me now

kberg commented on src/cards/Asteroid.ts:
```diff
@@ -17,18 +20,34 @@ export class Asteroid implements IProjectCard {
-        return player.canAfford(player.getCardCost(game, this) + REDS_RULING_POLICY_COST, game, false, true);
-      }
+        const temperatureMaxed = game.getVenusScaleLevel() === MAX_TEMPERATURE;
+        if (PartyHooks.shouldApplyPolicy(game, PartyName.REDS) && !temperatureMaxed) {
```
As I mentioned to @Lynesth, this kind of reformatting makes reviews harder. This is only 55 files, Lynesth's is over 200! :D

chosta replied:
It's going well, just a couple of merge conflicts :)

kberg commented on src/cards/IProjectCard.ts:
```diff
@@ -4,6 +4,7 @@ import { Player } from "../Player";
 import { Game } from "../Game";
 import { ResourceType } from "../ResourceType";
 import { Resources } from "../Resources";
+// import { CardMetadata } from "../cards/CardMetadata";
```
Remove?

chosta replied:
yes

kberg commented on src/cards/RequirementType.ts:
```diff
@@ -0,0 +1,17 @@
+    OCEANS = "Ocean",
+    VENUS = "Venus",
+    TR = "TR",
+    RESOURCE_TYPES = "Resource type",
```
"Resource types"? (The s at the end.)

chosta replied:
I'm not sure how to approach this. I follow the logic of OCEANS where we can have 1 Ocean or many Oceans and use it when creating the label text /toPlural/. Changing this to `Resource types` will result in failing unit tests. We can have a card with 1 Resource type as a requirement so I'd leave it as it.

kberg commented on src/components/card/CardDescription.ts:
```diff
@@ -0,0 +1,13 @@
+import Vue from "vue";
+//#TODO implement points per tag
```
Recommend TODO(chosta): so it's obvious who's responsible.

kberg commented on src/components/card/CardProductionBoxComponent.ts:
```diff
@@ -0,0 +1,47 @@
+        },
+    },
+    mounted: function () {
+        console.log(this.data);
```
eh?

chosta replied:
Just for debugging purposes. Won't be seen in the merge (promise!)

kberg commented on src/styles/cards_v2.less:
```diff
@@ -0,0 +1,443 @@
+@import "./variables.less";
```
Tell me more about v2 - is this a temporary name?

chosta replied:
Its temporary. And its v2 beacuse the css will be refactored and improved significantly.

kberg replied:
:+1:

kberg commented on src/utils/utils.ts:
```diff
@@ -12,3 +12,4 @@ export const playerColorClass = (
 };
 
 export const range = (n: number): Array<Number> => Array.from(Array(n).keys());
+export const firstLetterUpperCase = (s: string): string => s.charAt(0).toUpperCase() + s.slice(1);
```
You're only using this in one place? Just inline it.

chosta replied:
You are correct about this. My reasoning is that this is a function I have in all my js projects and it's usually used multiple times. I'll move it as this is the correct thing to do.

kberg commented:
I didn't realize this was five days neglected - sorry!

kberg commented on src/cards/CardMetadata.ts:
```diff
@@ -0,0 +1,10 @@
+    description?: string;
+    requirements?: CardRequirements;
+    onPlay?: Array<CardRow>;
+    victoryPoints?: number; //TODO(chosta): class to handle points per tag and other special cases
```
number | (c: Card) => number ?

chosta replied:
The TODO is for the general functionality but yes, something like that

kberg commented:
This looks great. Is it ready to pull in?

kberg commented on src/cards/CardRow.ts:
```diff
@@ -0,0 +1,14 @@
+import { CardAction } from "./CardAction";
+import { CardSpecial } from "./CardSpecial";
+
+export class CardRow {
```
So, what_is_ a `CardRow` exactly? (You know what I'm going to ask for next. Mwa ha ha ha ha.)

chosta replied:
I know but I just really wanted to go for a human understandable API and had to cut some corners. Feel free to suggest any improvements but this to me feels extremely comfortable:
```
public metadata: CardMetadata = {
        description:
            "Raise temperature 2 steps and gain 4 titanium. Remove up to 4 Plants from any player.",
        cardNumber: "011",
        onPlay: [
            CardRow.add([CardBonus.temperature(2)]),
            CardRow.add([CardBonus.titanium(4)]),
            CardRow.add([CardBonus.plants(-4).any()]),
        ],
    };
```
OR
```
public metadata: CardMetadata = {
        description:
            "Add 2 Microbes to ANOTHER card. Increase your heat production 3 steps and your Plant productions 1 step.",
        cardNumber: "170",
        onPlay: [
            CardRow.add([
                  CardProductionBox.add([
                        [CardBonus.heat(3)]
                   ])
            ]),
            CardRow.add([CardBonus.microbes(2), CardSpecial.asterix()]),
        ],
    };
```

kberg replied:
I hear you. But tell me in plain words what a CardRow is, please.

chosta commented:
> This looks great. Is it ready to pull in?
I'm in the process of functionally describing almost every element that is expected to be on a card. That includes Actions and Effects (should come in the next commits).
I'd like to try doing some tests on the Vue components with the new vue-type-checking approach that @bafolts suggested.
Once we have that, we can merge so others can join and help with the rest of the cards. It takes around 5 mins to describe a card with the metadata AND it's fun :)

kberg commented on src/cards/AerobrakedAmmoniaAsteroid.ts:
```diff
@@ -1,41 +1,57 @@
+        }
+
+        return new SelectCard(
+            "Select card to add 2 microbes",
```
@chosta I'm sorry -- the reformatting makes it difficult to identify the meaningful changes. Can you please remove them?

chosta replied:
@kberg Sorry about that but I actually have autoformatting turned off by default and do most of the formatting manually now...I do agree that it is really hard to identifiy meaningful changes so I'll make sure formatting goes to a separate commit. The only meaningful change for the existing ICard components in the previous commits would be adding the `metadata` property. I don't touch anything else.
Still, if you think I have to do git surgery so it would be easier to review I can go back and do some patching

kberg replied:
Please. This is an 80-file PR.

chosta replied:
I'll do my best. Will also stop adding new cards after today. Will refactor the existing functionality + builder patterns and tests. We can have a merge and then another PR. I'm also trying to carefully select commits and changes so you have a better time reviewing.

chosta commented:
@kberg Ares cards getting some love :)
BEFORE
![bio-before](https://user-images.githubusercontent.com/836179/97475674-36ab2600-1956-11eb-85ba-433130ab6d8e.png)

AFTER
![bio-after](https://user-images.githubusercontent.com/836179/97475683-390d8000-1956-11eb-9541-414820b248d9.png)

chosta commented:
@kberg I managed to separate formatting in 9d7bdb4 , 1ae8ae7 and 36e2790. Sadly, the rebase actions I did may have resulted in a bigger mess. Code is up to date and everything runs but not sure what to do with these commits. Is this ok, or should I figure out a cleaner approach? Please, advise..

kberg commented on src/cards/Algae.ts:
```diff
@@ -1,22 +1,43 @@
+            "Requires 5 ocean tiles. Gain 1 Plant and increase your Plant production 2 steps.",
+        cardNumber: "047",
+        requirements: new CardRequirements([CardRequirement.oceans(5)]),
+        onPlay: [
```
what is the relationship between `onPlay` which reads to me like something that defines behavior, and `CardRow` and `CardProducitonBox`, which read to me like objects that describe visualization?

kberg commented on src/cards/AsteroidMiningConsortium.ts:
```diff
@@ -24,4 +31,20 @@ export class AsteroidMiningConsortium implements IProjectCard {
+            "Requires that you have titanium production. Decrease any titanium production 1 step and increase your own 1 step.",
+        cardNumber: "002",
+        requirements: new CardRequirements([
+            CardRequirement.production(Resources.TITANIUM, -1)
```
This should be 1, not -1.

Lynesth replied:
The -1 is to not print a number in front of the requirement.
I agree it doesn't look intuitive at all and that decision (printing number or not) should probably be a third boolean argument with a default value.

kberg commented on src/cards/BiomassCombustors.ts:
```diff
@@ -24,4 +31,20 @@ export class BiomassCombustors implements IProjectCard {
+        description:
+            "Requires 6% oxygen. Decrease any Plant production 1 step and increase your Energy production 2 steps.",
+        cardNumber: "183",
+        requirements: new CardRequirements([CardRequirement.oxygen(6)]),
```
How will plant production be documented?

Lynesth replied:
I think what @chosta is working on is only taking care of cards visually, not of their behavior.

kberg commented on src/cards/CardBonus.ts:
```diff
@@ -0,0 +1,256 @@
+export class CardBonusTurmoilSpecial extends CardBonus implements ICardBonusWithResourceType {
+    constructor(amount: number, public resourceType: RequirementType) {
+        super(CardBonusType.TURMOIL_SPECIAL, amount);
+        this.resourceType = resourceType;
```
Is this line necessary if it's declared with a scope in the constructor?

kberg commented on src/cards/CardBonusType.ts:
```diff
@@ -0,0 +1,10 @@
+export enum CardBonusType {
```
What is a card bonus type? There is no such thing in the rulebook. So I suggest some documentation for someone coming around.

kberg commented on src/cards/CardSpecial.ts:
```diff
@@ -0,0 +1,55 @@
+import { CardSpecialType } from "./CardSpecialType";
+import { ComponentSize } from "../components/card/ComponentSize";
+
+export class CardSpecial {
```
What is a CardSpecial? This isn't clear.

kberg commented on src/cards/CardSpecial.ts:
```diff
@@ -0,0 +1,55 @@
+import { CardSpecialType } from "./CardSpecialType";
+import { ComponentSize } from "../components/card/ComponentSize";
```
I do not recommend importing `components` into non-components packages.

kberg commented on src/cards/CauseAndEffect.ts:
```diff
@@ -0,0 +1,66 @@
+import { CardSpecial } from "./CardSpecial";
```
So, as I'm reading this, I think perhaps you ought to move all of these visualization-focused classes into, say, `src/cards/viz`. (Not for nothing, I think all the _actual_ cards in `src/cards/*` ought to move to `src/cards/base`.

kberg reviewed (changes requested):
This will be easier to work with if there weren't so many cards to begin with. I realize this is getting big and difficult, but this PR is definitely big and difficult.

Lynesth commented on src/cards/CardRequirements.ts:
```diff
@@ -0,0 +1,16 @@
+import { CardRequirement, PartyCardRequirement } from "./CardRequirement";
+export class CardRequirements {
```
Just a suggestion but you could put that class in the same file as CardRequirement and PartyCardRequirement since it's really related.

Lynesth commented on src/cards/Algae.ts:
```diff
@@ -1,22 +1,43 @@
+    public cardType = CardType.AUTOMATED;
     public canPlay(player: Player, game: Game): boolean {
-      return game.board.getOceansOnBoard() >= 5 - player.getRequirementsBonus(game);
+        return game.board.getOceansOnBoard() >= 5 - player.getRequirementsBonus(game);
```
I think what @chosta is working on is only taking care of cards visually, not of their behavior.

chosta commented:
First of all, thanks for taking the time to review the changes. I agree that this PR is becoming too big and maybe we should have merged earlier. Let me try and address most of your concerns in this comment:

#### Enums
I've tried to reuse existing enum types such as Tags, Resources to describe every possible visual output on a card without having to create a new enum. I can easily have a CardBonusType and put all the enum values inside, this would improve some of the clutter but we would have enum types with repeating values.

#### Metadata - visual vs. behavior
My intention all along was to be able to describe the cards visually using metadata. I understand `metadata` as something that describes an item and not its behavior. Hence the `CardRow`, `CardProductionBox`, `CardSpecial`, etc. I didn't want to rewrite every ICard component so this:
```
game.board.getOceansOnBoard() >= 5 - player.getRequirementsBonus(game);
```
doesn't get implemented by the `CardRequirement` class as its job is to create a `requirements label` on top of the card.

The purpose of this approach is to have an API that describes what is drawn on the card that in a human readable way
```
public metadata: CardMetadata = {
        description:
            "Raise temperature 1 step and gain 2 titanium. Remove up to 3 Plants from any player.",
        cardNumber: "009",
        onPlay: [
            CardRow.add([CardBonus.temperature(1)]),
            CardRow.add([CardBonus.titanium(2)]),
            CardRow.add([CardBonus.plants(-3).any()]),
        ],
    };
```
Lets imagine we don't have this approach and try to describe a card like Lichen. We would have something like:
```
onPlay: { type: "plants", amount: 1, production: true } 
```
this can be drawn on the card without any issues but consider a card like CulturalMetropolis (and many other) where we have a production box that has two rows [-1 production, + 3 MC], a city tile and two delegates. We can describe the behavior but how do we know that we need to have the production box the city tile and the delegates on the same row? We can do that of course but not without any conditional logic.
![cultural](https://user-images.githubusercontent.com/836179/97676933-171d1600-1a9a-11eb-8256-0f8f95e52db5.png)

So the question here is, do we care if metadata `onPlay` (could be renamed) property only takes care of the visual representations of elements on the card canvas or we want to also describe behaviour with it?
+ If we choose to describe behaviour, I'll get rid of Boxes and Rows and start dealing with the logic inside of cards. This though would make the task much more complicated than I imagined it and would require more time and effort.
+ If we stick to using metadata for visual representation, I'll refactor the code by addressing the concerns in your comments (like CardSpecial, Enums, Builder patterns, etc.) Should we choose this approach, we can have 95% of the cards described programmatically with the current functionality. The rest would be exceptions and corporation cards, which require some additional work.

So it's time to decide. Please @bafolts @Lynesth @nwai90 @alrusdi @kberg @vincentneko @ssimeonoff check this comment and let me know what you think. Thanks for taking the time to help with this once again!

Lynesth commented:
I think it needs to start somewhere anyway. Even if metadata is only used to build cards visually for now, it could be extended later on to take care of some aspect of the behavior of the card.
It would of course be very nice to have everything thought of beforehand and layered for futur changes but I don't hink that's happening without too much work.

Regarding the PR itself:
- You could probably just make a "functional" one, that brings in the new classes and all (`CardRow`, `CardRequirement`, etc) and then make smaller PRs that adds metadata to cards (I get why kberg was baffled seeing mine before).
- Files that are related to metadata could probably be put inside a `metadata` folder to keep things neat.
- `onPlay` needs to be renamed to something else, it is pretty misleading. Since it's a property that holds the icons on the card, maybe something like `iconography`, `representation` or even `content` would make more sense.

Those are just my opinions on this of course :)

chosta commented:
> Even if metadata is only used to build cards visually for now, it could be extended later on to take care of some aspect of the behavior of the card.

Yes, this is how I imagine it too.

> * You could probably just make a "functional" one, that brings in the new classes and all (`CardRow`, `CardRequirement`, etc) and then make smaller PRs that adds metadata to cards (I get why kberg was baffled seeing mine before).

I'll break up this PR in chunks for sure, starting with CardRequirement. It needs some more work. I wanted to have the big picture and took the hands on approach by selecting both "easy" and "complex" cards and trying to draw them. It got me all the answers I need but now I have to take a step back and do proper planning.

kberg commented:
What's next?

chosta commented:
#1867
